### ISSUE TITLE
How to automatically save the model during training and add Early Stop by callbacks?

### ISSUE THREAD

**Issue by paipai880429** [dmlc/mxnet#1201]
Hi, all. I am a green hand of mxnet coming from Keras community. Multiple GPUs and distributed fashion are the two main reasons for me to turn to mxnet. I find out that the documents of mxnet are not perfect, at least not as good as Keras or Lasagne. I have two questions. 

Suppose that I am training for an image classification problem:

```
    train_dataiter = mx.io.ImageRecordIter(
        path_imgrec        = "./train.rec",
        rand_crop          = True,
        rand_mirror        = True,
        shuffle            = True,
        data_shape         = (3,224,224),
        batch_size         = batch_size,
        preprocess_threads = 4,
        prefetch_buffer    = 4,
        )
    test_dataiter = mx.io.ImageRecordIter(
        path_imgrec        = "./val.rec",
        rand_crop          = False,
        rand_mirror        = False,
        data_shape         = (3,224,224),
        batch_size         = batch_size,
        preprocess_threads = 4,
        prefetch_buffer    = 4,
        )

    model = mx.model.FeedForward(ctx=mx.gpu(0), symbol=softmax, num_epoch=10, learning_rate=0.1,     momentum=0.9, wd=0.0001,initializer=mx.init.Uniform(0.07))
    model.fit(X=train_dataiter, eval_data=test_dataiter, batch_end_callback=mx.callback.Speedometer(8))
```

How to make the code to enable it to automatically save the model as well as the parameters during training process (the highest validation classification accuracy)? Besides, how to add the early stopping by epoch_callbacks?

Thank you for your time. Happy New Year!

**Comment by alirezabagheri** [comment]
I am also interested in hearing the suggested solution for early stopping + saving model.

If you only intend to use a single callback-function argument, it appears one (hacky) way to go about it would be adding functionality to model.py:_train_multi_device() to use of the return value of 'call()' @ model.py:289. 

A better way may be adding a completely separate argument in _train_multi_device() specifically for the purpose of early stopping, then use heldout metric get() value to determine whether to break.

**Comment by paipai880429** [comment]
@alirezabagheri Great! I would appreciate if you can contribute and share you code for Early Stropping and Saving Model.

**Comment by piiswrong** [comment]
What exactly do you want early stopping to do? You can checkpoint the model and exit the program in epoch_end callbacks.

**Comment by paipai880429** [comment]
@piiswrong I just wanna early stopping to exit if the loss does not consistently drop down for 20 epochs. And meantime, the program always saves the model and its parameters to disk with current lowest loss. If you know how to implement this, could you please show me some example codes, please?

**Comment by alirezabagheri** [comment]
@piiswrong one cannot directly use epoch_end callbacks because (1) you need to know the current metric on eval set and (2) need to maintain state variables for what was the best heldout so far and how many epochs have past since this best. You can get around this by storing state in 'aux_params' however.

@paipai880429 assuming you are still interested in a solution, here's one way to go about it:
1. In python/mxnet/model.py:fit(...) add argument for early-stopping (perhaps with a default value, e.g. something like 'early_stopping_epochs = 10').
2. Within "fit()" function, pass this early-stopping variable to sub-function '_train_multi_device' and update the header for '_train_multi_device' to accept it (like step 1).
3. Within '_train_multi_device', search for 'eval_metric.get()' used on the evaluation set. Use this value and the early-stopping value to implement the logic to figure out when to stop. Call mxnet.model.save_checkpoint() to save model, and return to terminate.

**Comment by piiswrong** [comment]
@alirezabagheri There is an easier way of doing it. locals() is already added to batchparam. It captures all the local variables from _train_multi_device. You can add it to epoch_end_callback call and access everything from there.

**Comment by alirezabagheri** [comment]
@piiswrong Thanks for the quick response. Okay, I see the code has already changed a lot since my pull. 

Even with the latest code, not sure if your suggestion would exactly be easier. If the framework supported an **eval_**epoch_end_callback with argument for locals(); then that would be the best in the sense that everything could be done in the supplied callback function.

**Comment by amw5g** [comment]
@paipai880429 Don't know if you ever got your early stopping work complete.  But I've done something similar based on the xGBoost early stopping rounds parameter as well as the evaluation logger example in the mxnet team's documents:
````
mx.callback.early.stop <- function (period, nrounds, logger = NULL, maximize=TRUE, verbose=FALSE) 
{
    function(iteration, nbatch, env) {
        if (nbatch%%period == 0 && !is.null(env$metric)) {
            result <- env$metric$get(env$train.metric)
            if (nbatch != 0) 
                cat(paste0("Batch [", nbatch, "] Train-", result$name, 
                  "=", result$value, "\n"))
            if (!is.null(logger)) {
                if (class(logger) != "mx.metric.logger") {
                  stop("Invalid mx.metric.logger.")
                }
                logger$train <- c(logger$train, result$value)
                if (!is.null(env$eval.metric)) {
                  result <- env$metric$get(env$eval.metric)
                  if (nbatch != 0) 
                    cat(paste0("Batch [", nbatch, "] Validation-", 
                      result$name, "=", result$value, "\n"))
                  logger$eval <- c(logger$eval, result$value)
					  if (iteration > nrounds) {
						if(maximize) {
							if(sum(result$value > tail(logger$eval,nrounds)) == 0 ) { 
								bst.round <- which.max(logger$eval)
								if (verbose) {
									print(paste0("Best round: ", bst.round, "; Best metric: ", max(logger$eval)))
									print(paste0("This round: ", result$value))
								}
								return(FALSE) 
							}  
						} else if(sum(result$value < tail(logger$eval,nrounds)) == 0 ) { 
								bst.round <- which.min(logger$eval)
								if (verbose) {
									print(paste0("Best round: ", bst.round, "; Best metric: ", min(logger$eval)))
									print(paste0("This round: ", result$value))
								}
								return(FALSE) 							
						} 
					}
				}
			}
        return(TRUE)
		}
	}
}
````

You can call it like so:
````
model <- mx.model.FeedForward.create(lenet, X=train.array, y=train.y,
    num.round=500, array.batch.size=bs,
    learning.rate=lr, momentum=0.9, wd=0.00001,
    eval.metric=mx.metric.logloss,
    epoch.end.callback= mx.callback.early.stop(5,10,logger, maximize=FALSE, verbose=TRUE),
    eval.data=list(data=val.array, label=val.y))
````
